library/src/layout/par.rs

```diff
@@ -1233,13 +1270,30 @@ fn commit(
         }
     }
 
-    // Determine how much to justify each space.
+    // Determine how much addtional space is needed.
     let fr = line.fr();
-    let mut justification = Abs::zero();
-    if remaining < Abs::zero() || (line.justify && fr.is_zero()) {
+    let mut justification_ratio = 0.0;
+    let mut extra_justification = Abs::zero();
```

Comment:
I don't really understand the purpose of `extra_justification`.

Reply:
We can insert spaces between Chinese characters to stretch the line, but that is not preferred and should be last resort. (See [Procedures for Inter-Character Space Expansion](https://www.w3.org/TR/clreq/#procedures_for_inter-character_space_expansion) from Chinese Layout Requirements)

Let's say in a Chinese and Western mixed texts, we need to stretch a line with 3 space character and 3 Chinese character (not on line end). Every space character has stretchability of 1/6em and Chinese characters have stretchability of 0.

If the remaining space is less than 1/2em, we will not insert spaces between characters. Instead, we will stretch the 3 spaces to fill the remaining space.
On the other hand, if the remaining space is greater than 1/2em, stretching the 3 spaces by 1/6em will still leave us with some leftover length. In this case, we will add (remain - 1/2em) / 6 spaces to each adjustable character, including Chinese characters and spaces.

Reply:
Comments added in 587cf04649dfd9018b99fcd54febc67c1fbd2e29